crates/ra_proc_macro/src/process.rs

```diff
@@ -0,0 +1,209 @@
+    ) -> Result<Subtree, ra_tt::ExpansionError> {
+        let task = ExpansionTask {
+            macro_body: subtree.clone(),
+            macro_name: derive_name.to_string(),
+            attributes: None,
+            lib: dylib_path.to_path_buf(),
+        };
+
+        let result: ExpansionResult = self.send_task(Request::ExpansionMacro(task))?;
+        Ok(result.expansion)
+    }
+
+    pub fn send_task<R>(&self, req: Request) -> Result<R, ra_tt::ExpansionError>
+    where
+        R: TryFrom<Response, Error = &'static str>,
+    {
+        let sender = match &self.inner {
+            None => return Err(ra_tt::ExpansionError::Unknown("No sender is found.".to_string())),
+            Some(it) => it,
+        };
+
+        let (result_tx, result_rx) = bounded(0);
+
+        sender.send(Task::Request { req: req.into(), result_tx }).map_err(|err| {
+            ra_tt::ExpansionError::Unknown(format!(
+                "Fail to send task in channel, reason : {:#?} ",
+                err
+            ))
+        })?;
+
+        let res = result_rx.recv().unwrap();
+        match res {
+            Response::Error(err) => {
+                return Err(ra_tt::ExpansionError::ExpansionError(err.message));
+            }
+            _ => Ok(res.try_into().map_err(|err| {
+                ra_tt::ExpansionError::Unknown(format!(
+                    "Fail to get response, reason : {:#?} ",
+                    err
+                ))
+            })?),
+        }
+    }
+}
+
+fn client_loop(task_rx: Receiver<Task>, mut process: Process) {
+    let (mut stdin, mut stdout) = match process.stdio() {
+        None => return,
+        Some(it) => it,
+    };
+
+    loop {
+        let task = match task_rx.recv() {
+            Ok(task) => task,
+            Err(_) => break,
+        };
+
+        let (req, result_tx) = match task {
+            Task::Request { req, result_tx } => (req, result_tx),
+            Task::Close => break,
+        };
+
+        let res = match send_request(&mut stdin, &mut stdout, req) {
+            Ok(res) => res,
+            Err(_err) => {
+                let res = Response::Error(ResponseError {
+                    code: ErrorCode::ServerErrorEnd,
+                    message: "Server closed".into(),
+                });
+                if result_tx.send(res.into()).is_err() {
+                    break;
+                }
+                // Restart the process
+                if process.restart().is_err() {
+                    break;
+                }
+                let stdio = match process.stdio() {
+                    None => break,
+                    Some(it) => it,
+                };
+                stdin = stdio.0;
+                stdout = stdio.1;
+                continue;
+            }
+        };
+
+        if let Some(res) = res {
+            if result_tx.send(res).is_err() {
+                break;
+            }
+        }
+    }
+
+    let _ = process.child.kill();
```

Comment:
Let's implement `Drop` for process which kills the child?

Reply:
Fixed